node_modules/col-activities/lib/notifications/weekly.js

```diff
@@ -0,0 +1,583 @@
+var UserConstants = require('col-users/lib/constants');
+
+ /**
+ * Send weekly notification emails for all courses.
+ *
+ * @param  {Function}     callback    Standard callback function
+ */
+var collect = module.exports.collect = function(callback) {
+  var start = Date.now();
+
+  // Get all active courses.
+  CourseAPI.getCourses(function(err, courses) {
+    if (err) {
+      return log.error({'err': err}, 'Unable to retrieve courses for weekly notifications.');
+    }
+
+    // Iterate through all courses.
+    async.eachSeries(courses, collectCourse, function(err) {
+      if (err) {
+        return log.error({'err': err}, 'Unable to collect a course for weekly notifications.');
+      }
+      log.info({'duration': (Date.now() - start)}, 'Sent weekly notifications for all courses.');
+
+      return callback();
+    });
+  });
+};
+
+/**
+ * Send weekly notification emails for a given course.
+ *
+ * Note that this method is only made publicly available so it can be called from the integration
+ * tests. Do not call this method directly.
+ *
+ * @param  {Course}         course          The course for which to send weekly notifications
+ * @param  {Function}       callback        Standard callback function
+ * @api private
+ */
+var collectCourse = module.exports.collectCourse = function(course, callback) {
+  var start = Date.now();
+
+  // 1. Get users and activity data for the course.
+  getCourseData(course, function(err, courseData) {
+    // Don't pass errors up.
+    if (err) {
+      return callback();
+    }
+
+    // Return early if no activity data was found for the course.
+    if (!courseData) {
+      return callback();
+    }
+
+    // 2. Iterate through users in the course and send notifications to each.
+    var userIterator = handleUser.bind(null, course, courseData.activitySummary);
+    async.eachSeries(courseData.users, userIterator, function() {
+      log.info({
+        'course': course.id,
+        'duration': (Date.now() - start)
+      }, 'Sent weekly notifications for course.');
+
+      return callback();
+    });
+  });
+};
+
+/**
+ * Send notifications for a user in a course.
+ *
+ * Note that this function does not pass up errors.
+ *
+ * @param  {Course}     course              The course for which to send notifications
+ * @param  {Object}     activitySummary     Weekly activity data for the course
+ * @param  {User}       user                The user to notify
+ * @param  {Function}   callback            Standard callback function
+ * @api private
+ */
+var handleUser = function(course, activitySummary, user, callback) {
+  // Ignore users without an email address.
+  if (!user.canvas_email) {
+    return callback();
+  }
+
+  // Parse out activity data specific to this user.
+  // TODO: Include ranks for the user: last week's and this week's both.
+
+  var userData = activitySummary.users[user.id] || {};
+  if (!_.isEmpty(userData)) {
+    var topAsset = sampleMaximum(userData.assets, function(asset) {
```

Comment:
Can you indicate that we're trying to find the user's most popular asset during the last week ?